src/modules/Game.class.js

```diff
@@ -1,68 +1,271 @@
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  static Status = {
+    idle: 'idle',
+    playing: 'playing',
+    win: 'win',
+    lose: 'lose',
+  };
+
+  constructor(
+    initialState = [
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ],
+  ) {
+    this.initialState = initialState;
+    this.state = this.initialState.map((row) => [...row]);
+    this.status = Game.Status.idle;
+    this.score = 0;
+  }
+
+  moveLeft() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const newState = this.state.map((row) => this.move(row));
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(newState)) {
+      this.updateGameState(newState);
+      this.insertTiles();
+    }
+  }
+
+  moveRight() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const reversedState = this.state.map((row) => [...row].reverse());
+
+    const newState = reversedState.map((row) => {
+      return this.move(row).reverse();
+    });
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(newState)) {
+      this.updateGameState(newState);
+      this.insertTiles();
+    }
+  }
+
+  moveUp() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const rotateState = this.rotateLeft(this.getState());
+
+    const newState = rotateState.map((row) => this.move(row));
+
+    const unRotateState = this.rotateRight(newState);
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(unRotateState)) {
+      this.updateGameState(unRotateState);
+      this.insertTiles();
+    }
+  }
+
+  moveDown() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const rotateState = this.rotateRight(this.getState());
+
+    const newState = rotateState.map((row) => this.move(row));
+
+    const unRotateState = this.rotateLeft(newState);
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(unRotateState)) {
+      this.updateGameState(unRotateState);
+      this.insertTiles();
+    }
+  }
+
+  move(row) {
+    const newRow = [];
+    let i = 0;
+
+    while (i < row.length) {
+      const current = row[i];
+
+      if (current) {
+        let merged = false;
+
+        for (let j = i + 1; j < row.length; j++) {
+          const next = row[j];
+
+          if (next === current) {
+            newRow.push(current * 2);
+            this.updateGameScore(current * 2);
+            merged = true;
+            i = j + 1;
```

Comment:
When a merge occurs, you should skip over the next tile to prevent it from being merged multiple times in a single move. This logic might not handle consecutive merges correctly.